client/pluginmanager/csimanager/volume.go

```diff
@@ -453,12 +415,139 @@ func (v *volumeManager) ExpandVolume(ctx context.Context, volNS, volID, remoteID
+	hostStagingPath := v.stagingDirForVolume(v.mountRoot, vol.Namespace, vol.ID, usage)
+
+	// Make the staging path, owned by the Nomad User
+	if err := os.MkdirAll(hostStagingPath, 0700); err != nil && !os.IsExist(err) {
+		return "", false, fmt.Errorf("failed to create staging directory for volume (%s): %v", vol.ID, err)
+
+	}
+
+	// Validate that it is not already a mount point
 	m := mount.New()
-	isNotMount, err := m.IsNotAMountPoint(mountInfo.Source)
-	return !isNotMount, err
+	isNotMount, err := m.IsNotAMountPoint(hostStagingPath)
+	if err != nil {
+		return "", false, fmt.Errorf("mount point detection failed for volume (%s): %v", vol.ID, err)
+	}
+
+	return hostStagingPath, !isNotMount, nil
+}
+
+// ensureAllocDir attempts to create a directory for use when publishing a volume
+// and then validates that the path is not already a mount point (e.g when reattaching
+// to existing allocs).
+//
+// Returns whether the directory is a pre-existing mountpoint, the publish path,
+// and any errors that occurred.
+func (v *volumeManager) ensureAllocDir(vol *structs.CSIVolume, alloc *structs.Allocation, usage *UsageOptions) (string, bool, error) {
+	allocPath := v.allocDirForVolume(v.mountRoot, vol.ID, alloc.ID)
+
+	// Make the alloc path, owned by the Nomad User
+	if err := os.MkdirAll(allocPath, 0700); err != nil && !os.IsExist(err) {
+		return "", false, fmt.Errorf("failed to create allocation directory for volume (%s): %v", vol.ID, err)
+	}
+
+	// Validate that the target is not already a mount point
+	targetPath := v.targetForVolume(v.mountRoot, vol.ID, alloc.ID, usage)
+
+	m := mount.New()
+	isNotMount, err := m.IsNotAMountPoint(targetPath)
+
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		// ignore; path does not exist and as such is not a mount
+	case err != nil:
+		return "", false, fmt.Errorf("mount point detection failed for volume (%s): %v", vol.ID, err)
+	}
+
+	return targetPath, !isNotMount, nil
+}
+
+func volumeCapability(vol *structs.CSIVolume, usage *UsageOptions) (*csi.VolumeCapability, error) {
+	var opts *structs.CSIMountOptions
+	if vol.MountOptions == nil {
+		opts = usage.MountOptions
+	} else {
+		opts = vol.MountOptions.Copy()
+		opts.Merge(usage.MountOptions)
+	}
+
+	capability, err := csi.VolumeCapabilityFromStructs(usage.AttachmentMode, usage.AccessMode, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return capability, nil
+}
+
+func combineErrors(maybeErrs ...error) error {
+	var result *multierror.Error
+	for _, err := range maybeErrs {
+		if err == nil {
+			continue
+		}
+
+		result = multierror.Append(result, err)
+	}
+
+	return result.ErrorOrNil()
+}
+
+// serializedOp ensures that we only have one in-flight request per volume, and
+// keeps multi-step operations (ex. stage + publish) together in a single batch
+// rather than potentially interleaving
+func (v *volumeManager) serializedOp(ctx context.Context, volumeNS, volumeID string, fn func() error) error {
```

Comment:
I almost want to say that `fn` should take a context, too, `fn func(context.Context) error`, to clearly suggest that it should handle a shutdown signal, but I suppose you're covering that up within each closure, using the same context to call this func and within the closure body.

so it's 6 of one, half a dozen of the other, but I thought it, so I typed it. 😋